### PR TITLE
Add pending set to DAML triggers

### PR DESCRIPTION
**Pull request by cocreature**
### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: triggers/daml/Daml/Trigger.daml [hunk 1]
```diff
@@ -130,7 +144,7 @@ dedupExerciseByKey k c = do
   -- map to make these lookups cheaper.
   let cmds = concat $ map snd (Map.toList aState.commandsInFlight) <> map commands aState.emittedCommands
   unless (any ((Some (k, c) ==) . fromExerciseByKey @t) cmds) $
-    void $ emitCommands [exerciseByKeyCmd @t k c]
+    void $ emitCommands [exerciseByKeyCmd @t k c] []
```
Comment (aherrmann-da) [inline triggers/daml/Daml/Trigger.daml]: Why does `dedupExerciseByKey` behave differently from `dedupExercise` w.r.t. pening-set?

File: triggers/daml/Daml/Trigger.daml [hunk 2]
```diff
@@ -98,10 +110,12 @@ dedupCreate t = do
   -- map to make these lookups cheaper.
   let cmds = concat $ map snd (Map.toList aState.commandsInFlight) <> map commands aState.emittedCommands
   unless (any ((Some t ==) . fromCreate) cmds) $
-    void $ emitCommands [createCmd t]
+    void $ emitCommands [createCmd t] []
 
 -- | Exercise the choice on the given contract if it is not already
 -- in flight.
+-- dedupExercise will also mark the contract id as pending so it will not show up
+-- in `getContracts` in the next run.
```
Comment (aherrmann-da) [inline triggers/daml/Daml/Trigger.daml]: I'm not sure I understand the motivation behind this. Is this to model consuming choices? If so, does this interact with nonconsuming choices?

File: triggers/daml/Daml/Trigger.daml [hunk 3]
```diff
@@ -216,6 +232,7 @@ data TriggerAState = TriggerAState
   -- This will be used for dedupCreateCmd/dedupExerciseCmd helpers.
   , emittedCommands : [Commands]
   -- ^ Emitted commands in reverse because I can’t be bothered to implement a dlist.
+  , pendingContracts : Map CommandId [AnyContractId]
   , nextCommandId : Int
```
Comment (aherrmann-da) [inline triggers/daml/Daml/Trigger.daml]: Maybe add some haddocks for completeness?

File: triggers/daml/Daml/Trigger.daml (hunk 2 above)
Comment (cocreature) [reply triggers/daml/Daml/Trigger.daml]: Good point, I’ve removed the automatic pending thing here and then it behaves like `exerciseByKeyCmd`.

File: triggers/daml/Daml/Trigger.daml (hunk 1 above)
Comment (cocreature) [reply triggers/daml/Daml/Trigger.daml]: See comment below, they now both behave the same.

File: triggers/daml/Daml/Trigger.daml (hunk 3 above)
Comment (cocreature) [reply triggers/daml/Daml/Trigger.daml]: Good call, added.

**Pull request merged**